sbi/utils/score_utils.py

```diff
@@ -0,0 +1,597 @@
+        m: Scaling factor.
+        s: Standard deviation of the noise.
+        x_t: Observed data.
+        validate_args: Whether to validate arguments. Defaults to None.
+    """
+
+    arg_constraints = {
+        "low": constraints.real,
+        "high": constraints.real,
+        "m": constraints.real,
+        "s": constraints.positive,
+        "x_t": constraints.real,
+    }  # type: ignore
+    has_rsample: bool = False
+
+    def __init__(self, low, high, m, s, x_t, validate_args=None):
+        self.low = torch.as_tensor(low)
+        self.high = torch.as_tensor(high)
+        self.m = torch.as_tensor(m)
+        self.s = torch.as_tensor(s)
+        self.x_t = torch.as_tensor(x_t)
+
+        # Posterior is a truncated normal with:
+        self.mu = self.x_t / self.m
+        self.sigma = self.s / torch.abs(self.m)
+
+        # Standard Normal Distribution
+        self.standard_normal = Normal(torch.tensor(0.0, device=self.sigma.device), 1.0)
+
+        # Standardized truncation limits
+        self.alpha = (self.low - self.mu) / self.sigma
+        self.beta = (self.high - self.mu) / self.sigma
+
+        # Compute normalization constant Z
+        self.a = self.standard_normal.cdf(self.alpha)
+        self.b = self.standard_normal.cdf(self.beta)
+        self.Z = torch.clamp(self.b - self.a, min=1e-8)  # Avoid division by zero
+
+        batch_shape = torch.broadcast_shapes(
+            self.low.shape, self.high.shape, self.m.shape, self.s.shape, self.x_t.shape
+        )
+        event_shape = torch.Size()
+        super().__init__(batch_shape, event_shape, validate_args=validate_args)
+
+    @property
+    def mean(self):
+        # Standard normal pdf values
+        phi_alpha = torch.exp(self.standard_normal.log_prob(self.alpha))
+        phi_beta = torch.exp(self.standard_normal.log_prob(self.beta))
+
+        return self.mu + self.sigma * (phi_alpha - phi_beta) / self.Z
+
+    @property
+    def variance(self):
+        phi_alpha = torch.exp(self.standard_normal.log_prob(self.alpha))
+        phi_beta = torch.exp(self.standard_normal.log_prob(self.beta))
+
+        term = (phi_alpha - phi_beta) / self.Z
+        variance_adjustment = (
+            1 + (self.alpha * phi_alpha - self.beta * phi_beta) / self.Z - term**2
+        )
+
+        return self.sigma**2 * variance_adjustment
+
+    def sample(self, sample_shape=torch.Size()):
+        # Inverse CDF sampling for a truncated normal
+        u = torch.rand(sample_shape, device=self.low.device) * self.Z + self.a
+        sample = self.mu + self.sigma * self.standard_normal.icdf(
+            u
+        )  # Using icdf for stability
+        return torch.clamp(sample, min=self.low, max=self.high)
+
+    def log_prob(self, x):
+        lp = Normal(self.mu, self.sigma).log_prob(x) - torch.log(self.Z)
+        return torch.where((x >= self.low) & (x <= self.high), lp, -torch.inf)
+
+
+class UniformNormalConvolution(Distribution):
```

Comment:
same as above, it's great to have this! maybe we can re-use this more generally, e.g,. for the tests? 

Please add types.